src/thmap.c

```diff
@@ -470,28 +487,29 @@ find_edge_node(const thmap_t *thmap, thmap_query_t *query,
 
 	ASSERT(query->level == 0);
 
-	root_slot = atomic_load_explicit(&thmap->root[query->rslot],
-	    memory_order_relaxed);
+	/* Consume from prior release in root_try_put. */
+	root_slot = atomic_load_consume(&thmap->root[query->rslot]);
```

Comment:
@riastradh: Hmm.. I think this should be acquire fence rather than consume. Can you explain why are you using the weaker fence?

Reply:
All loads that must happen after the synchronizing store-release/load-consume are at addresses that are unconditional functions of the value root_slot returned by atomic_load_consume:

```
	// find_edge_node
	parent = THMAP_NODE(thmap, root_slot); // function of root_slot
	...
	node = atomic_load_consume(&parent->slots[off]);
	...
	if (atomic_load_relaxed(&parent->state) & ...)

	// thmap_get
	parent = find_edge_node(...);
	...
	leaf = get_leaf(thmap, parent, slot); // load-consumes parent->slots[slot]
	...
	if (!key_cmp_p(thmap, leaf, key, len)) // loads leaf->key (*)

	// find_edge_node_locked
	node = find_edge_node(...);
	lock_node(node); // accesses node->state
	...
	target = atomic_load_relaxed(&node->slots[*slot]);
	// &c.
```

We would only need load-acquire if, e.g., each branch had a bitmap of which child pointers are in use, and we required ordering between (a) updating the bitmap and (b) updating the child pointer.  But we always go from each parent node to exactly one child pointer in lookups.  (For updates we lock the parent node which implies an acquire operation anyway.)

(*) Exception: key_cmp_p should maybe use atomic_load_consume.  It's not actually a big deal because on Alpha, the membar implied by the first atomic_load_consume is good enough to cover this, and using atomic_load_consume would unnecessarily hurt performance; and on non-Alpha, the non-atomic leaf->key load is already a load-consume anyway.  But perhaps it would be clearer to write atomic_load_consume in key_cmp_p.

Reply:
I think in principle you are right.

However, I was not sure whether the dependency would be carried the given pointer slicing and dicing (see `THMAP_NODE()` and friends play with bits and operate on "base address/offset").  IIRC, there has been debate what exactly carriers a dependency (e.g. if I cast the pointer to different types, slice some bits, create aliases, etc -- does it carry?).  Looking at the "carries a dependency" definition in C17/C18 (section 5.1.2.4, paragraph 14), the second condition seems to cover this.

In the bigger picture, however, I am increasingly leaning towards a different view on `memory_order_consume`. Getting the memory ordering and synchronization right is already not an easy subject for many programmers.  If we add the tracking of pointer chains (or, perhaps, graphs.. or their leftover bits..) on top of that -- things become pretty complex for, I would argue, little benefit. Acquire fence is free on x86/amd64. There are some PowerPC and ARM offenders, but the current trends don't seem to be in favour of very weakly ordered architectures (oh, I really don't care about DEC Alpha).  So, I would advocate for a simplified model of using acquire/release fences, just ditching memory_order_consume.  Software will be just less buggy this way.  What do you think?